### PR TITLE
"tsc --watch should clear screen on new compilation" Two: Electric Boogaloo

### LINKED ISSUE #20379: Better visual distinction between --pretty --watch recompile results?
**TypeScript Version:**  2.7.0-dev.20171126

While running `tsc --pretty --watch`, there's not much visual distinction between different rounds of failures. Other than a few blank lines on the right half of the screen, they look very similar. When I have a lot of files I'm going through refactor errors through it can be hard to tell what's fixed or fresh.

image

For comparison, this is what TSLint's `stylish` formatter outputs...

image

...and this is what gulp outputs:

image
****

A few suggestions to make this display easier to read (replace "grey" with any non-white color):
* Use grey text for line/column numbers and/or grey+bold text for `TS2451`-style identifiers
* Use grey text for times
* Use grey and/or italic text for `File change detected.` lines
* In the `Compilation complete.` lines:
    * Say whether compilation failed or succeeded
    * Use green or red text appropriately
* Add a CLI option to clear the console when a new run starts

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
Fixes #20379

Port of #17756 since that hasn't been updated in a few weeks. The first iteration was written by @k0pernikus.

@mhegazy I like how the screen still has the "there's been a recompile" notice up top. Otherwise, for long recompiles, it can get a little confusing when the screen is completely blank for seconds at a time.

Re writing tests - I'm a little lost here. Is there somewhere this kind of test should go?
If it's acceptable to merge this PR without tests (which would be great for using the feature in nightlies!), I'd still be very interested in writing them later.

~~_Todo: look into 2 baseline unit test failures for .d.ts files_~~
~~_Todo: tests, if this doesn't get merged in first (which is fine by me!)_~~

### CODE REVIEW

File: src/compiler/watch.ts [hunk 1]
```diff
@@ -493,6 +493,10 @@ namespace ts {
         }
 
         function updateProgram() {
+            if (sys.clearScreen()) {
```
Comment (mhegazy) [inline src/compiler/watch.ts]: should be `if (sys.clearScreen)`

File: src/compiler/tsc.ts [hunk 2]
```diff
@@ -134,6 +134,10 @@ namespace ts {
     }
 
     function reportWatchModeWithoutSysSupport() {
+        if (sys.clearScreen()) {
```
Comment (mhegazy) [inline src/compiler/tsc.ts]: same here

Comment (mhegazy) [reply src/compiler/tsc.ts]: I do not see why you need this line here. checking in watch is sufficient.

File: src/compiler/watch.ts (hunk 1 above)
Comment (mhegazy) [reply src/compiler/watch.ts]: you should not use `sys` here. you should use ` watchingHost.system` instead.

**Comment by mhegazy** [comment]
you also need to accept the api baselines since you have changed the public API shape.

**Comment by mhegazy** [comment]
For the tests, all the tests for watch are in https://github.com/Microsoft/TypeScript/blob/master/src/harness/unittests/tscWatchMode.ts

**Comment by JoshuaKGoldberg** [comment]
After trying this out locally, I found it disruptive to have the screen immediately clear before the first "compilation succeeded" message. Removing that now and will add it to test cases.

**Comment by mhegazy** [comment]
thanks @JoshuaKGoldberg for picking this up and driving it through.

**Comment by dwieeb** [comment]
This is awesome, but _really_ screws with output when multiple `tsc` invocations are running concurrently (say, in a monorepo setup). I'm not seeing a way to disable it in the code merged from this PR. Should I create an issue?

**Comment by JoshuaKGoldberg** [comment]
@dwieeb see https://github.com/Microsoft/TypeScript/issues/21295

**Pull request merged**